### DISCUSSION TITLE
Accessing direct asset still works with Storage Asset Presets on Presets only

### DISCUSSION THREAD

**Discussion by flayks** [directus/directus#8887] [Feature Requests]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

With the Storage Asset Presets setting set on Presets only, we can still access an URL like `https://api.site.tld/assets/5dc1ea10-233b-449f-8d6b-bee77af038fe`, allowing to see the original uploaded file. This could be an issue, especially for full size images that we would like to protect?

I know it is not the safest way to leave large size images on the internet, but at least that would be cool to disable this direct link without a `key` parameter in order to protect this "full size" file? If not, is there a way to restrict it on the front-end side with Caddy or Nginx?

Just an idea.

### To Reproduce

Just access an asset url like `api.site.tld/assets/5dc1ea10-233b-449f-8d6b-bee77af038fe` without any `key` parameter

### What version of Directus are you using?

9.0.0-rc.98

### What version of Node.js are you using?

14.18.0

### What database are you using?

MySQL

### What browser are you using?

Brave

### What operating system are you using?

macOS Big Sur (11)

### How are you deploying Directus?

Locally and Docker

**Comment by joselcvarela** [comment]
Hello @flayks 
I cannot reproduce your problem. Didn't you configured any permissions for the public role?

**Comment by flayks** [comment]
> Hello @flayks I cannot reproduce your problem. Didn't you configured any permissions for the public role?

Not particularly, I just use the App Access Minimum preset for the Directus collections:

![image](https://user-images.githubusercontent.com/273716/137639029-3e7737c3-621b-4e74-b7aa-3167d7b16fe7.png)

**Comment by flayks** [comment]
@joselcvarela After some exploration, I still can access the original asset by removing the `key` parameter with an `access_token` in the asset URL 🤔, either with a Custom role or the default Public one

**Reply by azrikahar** [reply]
_moved this to discussion as I think it might be more appropriate here_

I believe there may be a misinterpretation here. Setting `presets` only for Storage Assets Transform will only restrict the ability of using transform parameters for fine grained control, as described here: https://docs.directus.io/reference/api/assets/#requesting-a-thumbnail

In the current codebase, an invalid query exception will thrown if the query is using anything _except_ the presets:

https://github.com/directus/directus/blob/6db9e0cd7dc2fac93964a6ac45e47fadf4d09eff/api/src/controllers/assets.ts#L129-L132

In any case, the Storage Asset Transform setting doesn't involve restriction of actual assets access. Hopefully @rijkvanzanten can correct me if I'm wrong here. 

---

However I do believe what you are looking for still can be a valid use case, hence the move to discussion 👍

**Reply by rijkvanzanten** [reply]
You're correct @azrikahar! This setting does not control whether or not you have access to the original file

**Comment by flayks** [comment]
Sorry for the unnecessary issue @azrikahar.

I get the logic here @rijkvanzanten. Although, I believe if we choose the option to limit assets generation to predefined presets, the original file should not easily be accessible (in this case a higher resolution photo).  
Would that makes sense to have a setting like "Disable original assets URL" or something to complete the Storage Asset Presets to stop accessing original assets without any key?

**Reply by benhaynes** [reply]
Right now we have `Storage Asset Transform` with the options:

* All
* None
* Presets Only

But I suggest we add a way to control the master file. I don't think permissions will work here, since that will likely disable transforms too. Perhaps this is a Project Setting (similar to the Transforms one above) that controls the master file access. 🤔  @rijkvanzanten ?

**Reply by rijkvanzanten** [reply]
> I believe if we choose the option to limit assets generation to predefined presets, the original file should not easily accessible (in this case a higher resolution photo).

_the option to limit assets generation_ says it all 🙂 The original file isn't a generated asset.

> But I suggest we add a way to control the master file. I don't think permissions will work here, since that will likely disable transforms too. Perhaps this is a Project Setting (similar to the Transforms one above) that controls the master file access. 🤔 @rijkvanzanten ?

Agreed. Permissions controls the whole file, so we need a separate flag to control "allow access to main file". However, that will break a couple things in the app, as we rely on that full path for things like the "download file" button etc

**Reply by flayks** [reply]
This is what I was thinking about indeed, like another settings coming along with the Storage Asset Transform (or not! could be set independently for more control) 👌

**Reply by sambitraze** [reply]
how can i restrict /assets/id to only authenticated access only

**Comment by br41nslug** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
